File: shmipc-preload/performance_analysis/ebpf/trace_shmipc_ebpf.py
#!/usr/bin/env python3
"""
eBPF User-space Program for shmipc Memory Tracing
Requires: bcc (BPF Compiler Collection)

Usage:
    sudo python3 trace_shmipc_ebpf.py <pid>
"""

from bcc import BPF
import sys
import time
import argparse
from collections import defaultdict

bpf_program = """
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>

struct event_t {
    u32 pid;
    u32 tid;
    u64 timestamp;
    u64 size;
    char comm[16];
    char func[32];
};

BPF_RINGBUF_OUTPUT(events, 1 << 24);
BPF_HASH(memcpy_stats, u32, u64);

int trace_memcpy(struct pt_regs *ctx) {
    struct event_t e = {};
    u64 id = bpf_get_current_pid_tgid();
    
    e.pid = id >> 32;
    e.tid = (u32)id;
    e.timestamp = bpf_ktime_get_ns();
    e.size = PT_REGS_PARM3(ctx);
    bpf_get_current_comm(&e.comm, sizeof(e.comm));
    __builtin_memcpy(e.func, "memcpy", 7);
    
    events.ringbuf_output(&e, sizeof(e), 0);
    
    u64 *count = memcpy_stats.lookup(&e.pid);
    if (count) {
        (*count) += e.size;
    } else {
        u64 init = e.size;
        memcpy_stats.update(&e.pid, &init);
    }
    
    return 0;
}

int trace_memmove(struct pt_regs *ctx) {
    struct event_t e = {};
    u64 id = bpf_get_current_pid_tgid();
    
    e.pid = id >> 32;
    e.tid = (u32)id;
    e.timestamp = bpf_ktime_get_ns();
    e.size = PT_REGS_PARM3(ctx);
    bpf_get_current_comm(&e.comm, sizeof(e.comm));
    __builtin_memcpy(e.func, "memmove", 8);
    
    events.ringbuf_output(&e, sizeof(e), 0);
    
    return 0;
}
"""

def main():
    parser = argparse.ArgumentParser(description='Trace memory operations in shmipc')
    parser.add_argument('pid', type=int, nargs='?', help='PID to trace (optional, traces all if not specified)')
    parser.add_argument('--duration', type=int, default=10, help='Duration to trace (seconds)')
    args = parser.parse_args()
    
    print(f"Loading eBPF program...")
    b = BPF(text=bpf_program)
    
    print("Attaching probes...")
    b.attach_uprobe(name="c", sym="memcpy", fn_name="trace_memcpy")
    b.attach_uprobe(name="c", sym="memmove", fn_name="trace_memmove")
    
    stats = defaultdict(lambda: {'count': 0, 'total_size': 0, 'funcs': defaultdict(int)})
    
    def handle_event(cpu, data, size):
        event = b['events'].event(data)
        
        if args.pid and event.pid != args.pid:
            return
        
        stats[event.pid]['count'] += 1
        stats[event.pid]['total_size'] += event.size
        stats[event.pid]['funcs'][event.func.decode()] += 1
        
        if event.size > 4096:
            print(f"[{event.pid}:{event.tid}] {event.func.decode()}: {event.size} bytes ({event.comm.decode()})")
    
    b['events'].open_ring_buffer(handle_event)
    
    print(f"Tracing for {args.duration} seconds... (Ctrl+C to stop)")
    print("=" * 80)
    
    try:
        for i in range(args.duration):
            b.ring_buffer_poll(timeout=1000)
            if i > 0 and i % 5 == 0:
                print(f"\n--- Stats at {i}s ---")
                for pid, data in sorted(stats.items(), key=lambda x: x[1]['total_size'], reverse=True)[:5]:
                    print(f"  PID {pid}: {data['count']} calls, {data['total_size']/1024/1024:.2f} MB")
    except KeyboardInterrupt:
        pass
    
    print("\n" + "=" * 80)
    print("=== Final Statistics ===")
    print("=" * 80)
    
    for pid, data in sorted(stats.items(), key=lambda x: x[1]['total_size'], reverse=True):
        print(f"\nPID {pid}:")
        print(f"  Total calls: {data['count']}")
        print(f"  Total bytes: {data['total_size']:,} ({data['total_size']/1024/1024:.2f} MB)")
        print(f"  Avg size: {data['total_size']/data['count']:.0f} bytes")
        print(f"  By function:")
        for func, count in sorted(data['funcs'].items(), key=lambda x: x[1], reverse=True):
            print(f"    {func}: {count} calls")

if __name__ == '__main__':
    main()

File: shmipc-preload/performance_analysis/ebpf/trace_shmipc_ebpf.c
// trace_shmipc_ebpf.c - eBPF 程序：追踪 shmipc 内存拷贝
// 编译: clang -g -O2 -target bpf -D__TARGET_ARCH_x86 -c trace_shmipc_ebpf.c -o trace_shmipc_ebpf.o

#include "vmlinux.h"
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>

char LICENSE[] SEC("license") = "GPL";

struct event_t {
    u32 pid;
    u32 tid;
    u64 timestamp;
    u64 size;
    char comm[16];
    char func[32];
};

struct {
    __uint(type, BPF_MAP_TYPE_RINGBUF);
    __uint(max_entries, 1 << 24);
} events SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, 10240);
    __type(key, u32);
    __type(value, u64);
} memcpy_stats SEC(".maps");

SEC("uprobe/libc.so.6:memcpy")
int BPF_UPROBE(trace_memcpy, void *dst, void *src, size_t n) {
    struct event_t *e;
    
    e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
    if (!e) return 0;
    
    u64 id = bpf_get_current_pid_tgid();
    e->pid = id >> 32;
    e->tid = (u32)id;
    e->timestamp = bpf_ktime_get_ns();
    e->size = n;
    bpf_get_current_comm(&e->comm, sizeof(e->comm));
    __builtin_memcpy(e->func, "memcpy", 7);
    
    bpf_ringbuf_submit(e, 0);
    
    u32 pid = e->pid;
    u64 *count = bpf_map_lookup_elem(&memcpy_stats, &pid);
    if (count) {
        __sync_fetch_and_add(count, n);
    } else {
        u64 init = n;
        bpf_map_update_elem(&memcpy_stats, &pid, &init, BPF_ANY);
    }
    
    return 0;
}

SEC("uprobe/libc.so.6:memmove")
int BPF_UPROBE(trace_memmove, void *dst, void *src, size_t n) {
    struct event_t *e;
    
    e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
    if (!e) return 0;
    
    u64 id = bpf_get_current_pid_tgid();
    e->pid = id >> 32;
    e->tid = (u32)id;
    e->timestamp = bpf_ktime_get_ns();
    e->size = n;
    bpf_get_current_comm(&e->comm, sizeof(e->comm));
    __builtin_memcpy(e->func, "memmove", 8);
    
    bpf_ringbuf_submit(e, 0);
    
    return 0;
}

SEC("uprobe/libc.so.6:memset")
int BPF_UPROBE(trace_memset, void *s, int c, size_t n) {
    struct event_t *e;
    
    e = bpf_ringbuf_reserve(&events, sizeof(*e), 0);
    if (!e) return 0;
    
    u64 id = bpf_get_current_pid_tgid();
    e->pid = id >> 32;
    e->tid = (u32)id;
    e->timestamp = bpf_ktime_get_ns();
    e->size = n;
    bpf_get_current_comm(&e->comm, sizeof(e->comm));
    __builtin_memcpy(e->func, "memset", 7);
    
    bpf_ringbuf_submit(e, 0);
    
    return 0;
}
